### PR TITLE
Initial pull request

### PR DESCRIPTION
**Pull request by VeljkoDeljanin**

### CODE REVIEW

File: src/client/game_client.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    def __send_data(self, action: Action, data: dict = None) -> dict:
+        if data is not None:
+            msg = action.value.to_bytes(4, "little") + self.__encode_message(data)
+        else:
+            msg = action.value.to_bytes(8, "little")
+
+        self.__socket.sendall(msg)
+
+        response_msg = self.__socket.recv(MAX_MESSAGE_SIZE)
```
Comment (irina-scherbitskaya) [inline src/client/game_client.py]: Socket.recv can return less data than MAX_MESSAGE_SIZE, you need to subtract data in chunks of 1024/2048 in a loop with a check that all the data was actually subtracted. For this you need to first extract only action code and msg length

File: src/game.py [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    def add_player(self, name: str, password: str = None, is_observer: bool = None) -> None:
+        if self.__game_players >= self.__max_players:
+            is_observer = True
+
+        if not is_observer:
+            self.__game_players += 1
+
+        self.__all_players += 1
+
+        player: Player
```
Comment (irina-scherbitskaya) [inline src/game.py]: it is not customary to do in python, better to specify the type annotation with the initialization

File: src/game.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    def __update_turn(self) -> None:
+        game_state = self.__current_client.game_state()
+
+        self.current_turn = game_state["current_turn"]
+        self.__current_player_idx = game_state["current_player_idx"]
+
+        print()
+        if self.__current_player_idx != 0:
+            self.__current_player = self.__players_in_game[self.__current_player_idx]
+            self.__current_client = self.__all_clients[self.__current_player]
```
Comment (irina-scherbitskaya) [inline src/game.py]: But if it is remote player?

File: src/client/game_client.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
+import socket
+import json
+
+from src.constants import Action, Result, SERVER_HOST, SERVER_PORT, MAX_MESSAGE_SIZE
+
+
+class Client:
```
Comment (irina-scherbitskaya) [inline src/client/game_client.py]: better — “ServerConnection”

File: src/game.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    self.running = False
+
+            self.map.draw_map(self.current_turn, self.num_turns, self.current_round, self.num_rounds)
+
+            self.__clock.tick(60)
+
+            for player in self.__players_in_game.values():
+                player.next_turn_sem.release()
```
Comment (irina-scherbitskaya) [inline src/game.py]: I don't really see the point of using semaphores here, every tick only 1 bot needs to make a full turn, and the action of the turn for the rest is not expensive

File: src/game.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+            player.round_reset()
+
+        self.map = Map(game_map, game_state, self.__players_in_game)
+
+        for player in self.__players_in_game.values():
+            player.round_update(self.map)
+
+    def __update_turn(self) -> None:
+        game_state = self.__current_client.game_state()
+
```
Comment (irina-scherbitskaya) [inline src/game.py:157]: but if there are not only your bots in the game, there was your turn and the turn has not yet been updated, will you make a turn a second time? maybe make a check that if the turn has already been processed, then skip

File: src/vehicles/tank.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    def get_player_id(self) -> int:
+        return self.__player_id
+
+    def get_sp(self) -> int:
+        return self.__sp
+
+    def get_position(self) -> Hex:
+        return self.__position
+
+    def get_spawn_position(self) -> Hex:
```
Comment (irina-scherbitskaya) [inline src/vehicles/tank.py]: can use decorator @ property

File: src/constants.py [hunk 8]
```diff
@@ -99,3 +99,12 @@ class Result(IntEnum):
 LEGEND_TEXT_POSITION = [(1005, 56), (1005, 93), (1005, 130), (1005, 167),
                         (1005, 204), (1005, 601), (1005, 638), (1005, 675)]
 LEGEND_NAME = ["spg", "light_tank", "heavy_tank", "medium_tank", "at_spg", "catapult", "heavy_repair", "light_repair"]
+
+# abs values of optimal hex coordinates for each tank
+optimal_hexes = {
```
Comment (irina-scherbitskaya) [inline src/constants.py]: upper case

File: src/logic.py [hunk 9]
```diff
@@ -1,43 +1,66 @@
...
-        visited = [h for h in visited if h not in self.__map.get_spawn()
-                   and h not in self.__map.get_tank_positions().values()]
+        visited = [h for h in visited if h not in self.__map.spawn
+                   and h not in self.__map.tank_positions.values()]
         visited.sort(key=lambda x: x)
 
         move_to = []
         if visited:
-            d = Hex.distance(Hex(0, 0, 0), visited[0])
+            if visited[0] == Hex(0, 0, 0):
```
Comment (irina-scherbitskaya) [inline src/logic.py]: d = Hex.distance(Hex(0, 0, 0), visited[1]) if visited[0] == Hex(0, 0, 0) else Hex.distance(Hex(0, 0, 0), visited[0])


and you can create  .isZero method

File: src/logic.py [hunk 10]
```diff
@@ -1,43 +1,66 @@
...
+                d = Hex.distance(Hex(0, 0, 0), visited[1])
```
Comment (irina-scherbitskaya) [inline src/logic.py]: Hex(0, 0, 0) often used, you can just put it in a variable

File: src/logic.py [hunk 11]
```diff
@@ -139,63 +164,122 @@ def __straight_trajectory(self, tank: Tank) -> tuple[Optional[Hex], Optional[lis
...
     def reset_shoot_actions(self, player_id: int) -> None:
-        self.__map.get_shoot_actions()[player_id] = []
+        self.__map.shoot_actions[player_id] = []
 
     def is_in_base(self, tank_pos: Hex) -> bool:
-        return tank_pos in self.__map.get_base()
+        return tank_pos in self.__map.base
 
     # Detects if the given coordinates are off the grid
     def __offTheGridDetection(self, h: Hex) -> bool:
```
Comment (irina-scherbitskaya) [inline src/logic.py]: why is the codestyle different here?
and __hexNeighbors

Comment (Zeljko103) [reply src/logic.py]: Will be fixed in no time. Initially, two of us where working on these methods, so I forgot to change it

**Pull request merged**